### DISCUSSION TITLE
Supporting More Sensors: We Want Your Feedback!

### DISCUSSION THREAD

**Discussion by dabdoue** [Moddable-OpenSource/moddable#1029] [Announcements]
A common Moddable SDK feature request is support for additional sensors. Building out new ECMA-419 conformant drivers for those sensors--and anticipating the sensor needs of potential Moddable SDK developers--is a priority of the Moddable team.
I started a [Popular Sensor Tracker](https://docs.google.com/spreadsheets/d/1J13KCMnz3On9mZkCT7aGd2btnhny-MZ81V5a3HCNEBs/edit?usp=sharing) to help prioritize our efforts and to provide a "hit list" for community contributions. The list is based on the sales data and popularity of sensors on sites like Amazon and SparkFun, as well as devices commonly supported by other projects.

My hope is that this list will be the start of a conversation. We want to hear from the Moddable community about what sensors you use, what components you would add to or drop from this list, and what categories we may have missed entirely. Please jump in this thread and let us know what sensors you would like to see supported!

**Comment by phoddie** [comment]
@Sineos – This is something of a complementary effort to your excellent "[wish list](https://github.com/Moddable-OpenSource/moddable/issues/999)". Should we try to merge these somehow?

**Reply by dabdoue** [reply]
I made sure to include the sensors that you had on your wish list @Sineos , as I thought they were great suggestions

**Reply by Sineos** [reply]
I think this an excellent idea. What might be a nice addition is the information if a given sensor is still in active production or already obsolete / replaced.

Of course implementing such sensors is entirely at the discretion of the moddable project but in my opinion a workable prioritization could be:
1. Actively requested
2. Popular and still in active production
3. Popular but already obsolete

Edit: Naturally unless provided by the community which would be the best option

**Reply by Sineos** [reply]
@dabdoue what is the meaning of the ECMA-419 column in your list?

**Reply by dabdoue** [reply]
@Sineos very sorry for the late reply. 

To address your first concern, I think the prioritization you mentioned makes sense. I did my best to ensure that all the sensors on this list are either still actively in production or being sold actively online. 

One the second, the meaning of that column is whether the sensor already conforms to a sensor category in the ECMA-419 standard.

**Reply by Sineos** [reply]
> One the second, the meaning of that column is whether the sensor already conforms to a sensor category in the ECMA-419 standard.

@dabdoue thanks for the reply. The ECMA-419 stuff is confusing me. According to my understanding only the sensors in https://github.com/Moddable-OpenSource/moddable/tree/public/modules/drivers/sensors are actually supported under this standard.

> I did my best to ensure that all the sensors on this list are either still actively in production or being sold actively online.

I'm not sure if "actively sold" is a good criteria if someones digs into creating something new. According to my assessment here https://github.com/Moddable-OpenSource/moddable/tree/public/modules/drivers#readme quite some of the already supported sensors are no longer recommended by the manufacturer for new design or are marked as obsolete / out of production.

This also can lead to counterfeit / refurbished / mislabled chips. One prominent example I encountered myself are the MPU-6050 chips. Insanely popular but a lot of rubbish in the market.

**Reply by dabdoue** [reply]
> The ECMA-419 stuff is confusing me. According to my understanding only the sensors in https://github.com/Moddable-OpenSource/moddable/tree/public/modules/drivers/sensors are actually supported under this standard.

The link you gave are the sensors that have support from the Moddable SDK, but the standard also defines what types of sensors would be considered compatible with the standard, thus defining a standard way to implement a sensor that is not directly supported by the SDK (ex: a temperature sensor is supported by the standard, but not a radon sensor)

> I'm not sure if "actively sold" is a good criteria if someones digs into creating something new. According to my assessment here https://github.com/Moddable-OpenSource/moddable/tree/public/modules/drivers#readme quite some of the already supported sensors are no longer recommended by the manufacturer for new design or are marked as obsolete / out of production.
This also can lead to counterfeit / refurbished / mislabled chips. One prominent example I encountered myself are the MPU-6050 chips. Insanely popular but a lot of rubbish in the market.

That is a valid point. I will say that the vast majority of the sensors on this list are described as "actively in production" on their product pages, and if that information wasn't directly available, then the sensor was being sold by a reputable vendor like Sparkfun, or by the manufacturer itself. 

I do think it is worth adding that information explicitly however, so I will add a column with that information for all the sensors.

**Reply by Sineos** [reply]
> The link you gave are the sensors that have support from the Moddable SDK, but the standard also defines what types of sensors would be considered compatible with the standard

Do you establish the list from the Moddable SDK perspective or more generally?
If it is from the Moddable SDK perspective then I think the information that a sensor is measuring an unit that is covered under a certain standard is of little added value as long as the sensor is not actually supported in the Moddable SDK.
Maybe differentiate between `Covered under ECMA-419` and `Implemented as ECMA-419`? Whereas I would consider the latter sufficient.

Edit:
In fact, today there are even 3 different states:
* `Covered under ECMA-419`: Measuring a unit that is implemented in the standard
* `Implemented as ECMA-419`: Moddable SDK implementation conforming to the standard
* `Implemented`: Moddable SDK implementation not conforming to the standard

**Reply by dabdoue** [reply]
I think you make a great point. When I go through the list updating for "in-production", I can also update this column.

Thank you so much for your input!

**Comment by mauroForlimpopoli** [comment]
Congratulation for this initiative !
I saw the sensors list and it is already very complete, but I can suggest AD7783. It's a 24-Bit Σ-Δ ADC used to drive PT100 sensors (with 3 wires) for very accurate temperature measuring. It works on SPI line
In next days I will check for other sensors
Good job

**Reply by phoddie** [reply]
Moddable SDK 3.9.1 contains a driver for the [Bosch BME680](https://www.bosch-sensortec.com/products/environmental-sensors/gas-sensors/bme680/) sensor. This is likely the most complex sensor software I've implemented to date. Bosch provides nearly [2000 lines of C code](https://github.com/Moddable-OpenSource/moddable/blob/public/modules/drivers/sensors/bme68x/boschlib/bme68x.c) to use the sensor. While it is possible to use the sensor without that, it is challenging. So, I [wrapped](https://github.com/Moddable-OpenSource/moddable/blob/public/modules/drivers/sensors/bme68x/bme68xlib.c) the C code from Bosch into a [BME68x JavaScript class](https://github.com/Moddable-OpenSource/moddable/blob/70faab72f45651ebc841f3e1072c0446b3db39a2/modules/drivers/sensors/bme68x/bme68x.js#L22-L101). The BME68x JavaScript API basically mirrors the Bosch API. Then I used the Bosch JavaScript class to implement the ECMA-419 sensor driver. That is [very compact](https://github.com/Moddable-OpenSource/moddable/blob/70faab72f45651ebc841f3e1072c0446b3db39a2/modules/drivers/sensors/bme68x/bme68x.js#L104-L158) -- just over 50 lines -- because all the hard work is done by the Bosch C code.

The Bosch C code comes with 4 examples to show how to use it. I ported all of those to JavaScript as well, so you can experiment with the different modes of operation of the sensor hardware: [forced](https://github.com/Moddable-OpenSource/moddable/blob/public/examples/drivers/sensors/bme68x/examples/forced_mode.js), [sequential](https://github.com/Moddable-OpenSource/moddable/blob/public/examples/drivers/sensors/bme68x/examples/sequential_mode.js), [parallel](https://github.com/Moddable-OpenSource/moddable/blob/public/examples/drivers/sensors/bme68x/examples/parallel_mode.js), and [self-test](https://github.com/Moddable-OpenSource/moddable/blob/public/examples/drivers/sensors/bme68x/examples/self_test.js). There's also an [example](https://github.com/Moddable-OpenSource/moddable/blob/public/examples/drivers/sensors/bme68x/examples/sensor.js) for using the ECMA-419 BME68x sensor driver.

Finally, for Node-RED MCU Edition users, the MCU Sensor node database has been updated to include the BME68x so you drop it into your flows.

![image](https://user-images.githubusercontent.com/6856458/236583029-956cdcd9-e992-4034-b696-acbac8918d93.png)
